### PR TITLE
gsoc: Add Chain Action

### PR DESCRIPTION
**Pull request by Nosferatul**

### CODE REVIEW

File: coalib/output/ConsoleInteraction.py [hunk 1]
```diff
@@ -636,26 +638,85 @@ def ask_for_action_and_apply(console_printer,
...
-        console_printer.print(
-            format_lines(chosen_action.SUCCESS_MESSAGE, symbol='['),
-            color=SUCCESS_COLOR)
-        failed_actions.discard(action_name)
-    except Exception as exception:  # pylint: disable=broad-except
-        logging.error('Failed to execute the action {} with error: {}.'.format(
-            action_name, exception))
-        failed_actions.add(action_name)
-    return True
+    if action_name == 'ChainPatchAction':
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Holy crap this is a really long sequence, Can you break this down to separate methods? Especially for the ones that are duplicates.

Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: It's more maintainable if you break it down into little methods than a huge sequence like this.

**Comment by Nosferatul** [comment]
i think it's better now @yukiisbored I fixed some of the bugs

File: coalib/output/ConsoleInteraction.py [hunk 2]
```diff
@@ -638,24 +744,33 @@ def ask_for_action_and_apply(console_printer,
     """
     action_name, section = print_actions(console_printer, section,
                                          metadata_list, failed_actions)
+
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Why is this needed?

Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: no, it's not.

Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: euhm, can you remove it then? It's a not related to the patch.

File: coalib/output/ConsoleInteraction.py [hunk 3]
```diff
@@ -607,6 +609,113 @@ def print_actions(console_printer, section, actions, failed_actions):
     return get_action_info(section, actions[choice - 1], failed_actions)
 
 
+def apply_action(action_name,
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: Shouldnt we have smt similar for apply action action?
Anyway, should have docstring

File: coalib/output/ConsoleInteraction.py [hunk 4]
```diff
@@ -607,6 +609,113 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+            if action_choice == '(N)':
+                return False
+            elif action_choice in actions_name_dict:
+                action_choice_made = action_choice
+                action_choice = actions_name_dict[action_choice]
+                flag = False
+                for action_key in action_dict:
+                    if action_choice == action_key:
+                        chosen_action = action_dict[action_key]
+                        flag = True
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: Are there more possibilities or can u just break?

File: coalib/output/ConsoleInteraction.py [hunk 5]
```diff
@@ -607,6 +609,113 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+                if flag:
+                    for i, action_de in enumerate(metadata_list, 1):
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: Rename var to action_desc if that is what its for

Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: I dont like i, use index if necessary

Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: Or a better var name

File: coalib/results/result_actions/ChainPatchAction.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+    def apply(self,
+              result,
+              original_file_dict,
+              file_diff_dict,
+              actions: str):
+        """
+        (C)hain actions
+
+        :param actions: A string of all actions
```
Comment (Adrianzatreanu) [inline coalib/results/result_actions/ChainPatchAction.py]: End line with period

**Comment by Adrianzatreanu** [comment]
Took a look and gave you some comments. @yukiisbored opinions?

File: coalib/output/ConsoleInteraction.py (hunk 4 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: right

File: coalib/output/ConsoleInteraction.py [hunk 7]
```diff
@@ -607,6 +609,139 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+    :param failed_actions:  A set of all actions that have failed. A failed
+                            action remains in the list until it is successfully
+                            executed.
+    :param result:          Result corresponding to the actions.
+    :param file_diff_dict:  If it is an action which applies a patch, this
+                            contains the diff of the patch to be applied to
+                            the file with filename as keys.
+    :param file_dict:       Dictionary with filename as keys and its contents
+                            as values.
+    :return:                Returns a boolean value. True will be returned, if
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: Not the best wording tbh. Make it mkre like: return true if there is the possibility of the user selecting bla bla, or false otherwise.

**Comment by Nosferatul** [comment]
done @Adrianzatreanu

**Comment by Nosferatul** [comment]
http://imgur.com/pmM3ZlJ

File: coalib/output/ConsoleInteraction.py [hunk 8]
```diff
@@ -607,6 +609,138 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+                        action_dict,
+                        failed_actions,
+                        result,
+                        file_diff_dict,
+                        file_dict):
+    """
+    Try to apply the given action.
+
+    :param action_name:     The name of the action.
+    :choose_action:         The action object that will be applied.
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: shouldnt this be :param..: ?

File: coalib/output/ConsoleInteraction.py [hunk 9]
```diff
@@ -607,6 +609,138 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+    :param console_printer: Object to print messages on the console.
+    :param section:         Currently active section.
+    :param metadata_list:   Contains metadata for all the actions.
+    :param action_dict:     Contains the action names as keys and their
+                            references as values.
+    :param failed_actions:  A set of all actions that have failed. A failed
+                            action remains in the list until it is successfully
+                            executed.
+    :param result:          Result corresponding to the actions.
+    :param file_diff_dict:  If it is an action which applies a patch, this
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: this description makes no sense to me

File: coalib/output/ConsoleInteraction.py [hunk 10]
```diff
@@ -607,6 +609,138 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+    try:
+        chosen_action.apply_from_section(result,
+                                         file_dict,
+                                         file_diff_dict,
+                                         section)
+        console_printer.print(
+            format_lines(chosen_action.SUCCESS_MESSAGE, symbol='['),
+            color=SUCCESS_COLOR)
+        failed_actions.discard(action_name)
+    except Exception as exception:  # pylint: disable=broad-except
```
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: what does this comment do?

File: coalib/output/ConsoleInteraction.py (hunk 9 above)
Comment (Adrianzatreanu) [inline coalib/output/ConsoleInteraction.py]: what does this exactly do?

File: coalib/output/ConsoleInteraction.py (hunk 10 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: the comment was before, i just made a new function to not duplicate code

File: coalib/output/ConsoleInteraction.py (hunk 9 above)
Comment (Nosferatul) [inline coalib/output/ConsoleInteraction.py]: it's a dictionary with differences for a file. This is the description copied from other function already implemented.

**Comment by Adrianzatreanu** [comment]
This looks great to me. @yukiisbored please take a look too, and ack it, as it has my ack.

File: coalib/output/ConsoleInteraction.py [hunk 11]
```diff
@@ -607,6 +609,138 @@ def print_actions(console_printer, section, actions, failed_actions):
...
+                            contains the diff of the patch to be applied to
+                            the file with filename as keys.
+    :param file_dict:       Dictionary with filename as keys and its contents
+                            as values.
+    :return:                Return False if the action that is applied is Do
+                            (N)othing or True otherwise.
+    """
+    flag = False
+
+    actions_name_dict = {'(A)': 'ApplyPatchAction', '(I)': 'IgnoreResultAction',
```
Comment (yukiisbored) [inline coalib/output/ConsoleInteraction.py]: Hmm, can you generate this dynamicly? Since they're declared in the docstrings.

**Comment by yukiisbored** [comment]
Looks good, just one little problem.

**Comment by Adrianzatreanu** [comment]
ack f1d4490

**Comment by yukiisbored** [comment]
@Nosferatul needs rebase.

**Comment by Adrianzatreanu** [comment]
i restarted your build and it passed :)

**Comment by Adrianzatreanu** [comment]
reack 45a7369

**Comment by Adrianzatreanu** [comment]
@rultor merge
